src/thunderbots/software/test/util/gradient_descent_optimizer.cpp

```diff
@@ -0,0 +1,106 @@
+/**
+ * Tests for the `GradientDescentOptimizer`
+ */
+
+#include <gtest/gtest.h>
+
+#include <cmath>
+
+#include "util/gradient_descent.h"
+
+using namespace Util;
+
+TEST(GradientDescentOptimizerTest, minimize_single_valued_function)
+{
+    GradientDescentOptimizer<1> gradientDescentOptimizer({0.1});
+
+    // f = x^2
+    auto f = [](std::array<double, 1> x) { return std::pow(x.at(0), 2.0); };
+
+    auto min = gradientDescentOptimizer.minimize(f, {1}, 100);
+
+    EXPECT_NEAR(min.at(0), 0, 0.1);
+}
+
+TEST(GradientDescentOptimizerTest, maximize_single_valued_function)
+{
+    GradientDescentOptimizer<1> gradientDescentOptimizer({0.1});
+
+    // f = -x^2
+    auto f = [](std::array<double, 1> x) { return -std::pow(x.at(0), 2.0); };
+
+    auto max = gradientDescentOptimizer.maximize(f, {1}, 100);
+
+    EXPECT_NEAR(max.at(0), 0, 0.1);
+}
+
+TEST(GradientDescentOptimizerTest, minimize_multi_valued_function)
+{
+    GradientDescentOptimizer<2> gradientDescentOptimizer({0.1, 0.1});
```

Comment:
Just a question here. In comments below you mention the purpose of the parameter weights are to normalize the parameters so they are roughly homogeneous.

```
// Weights used to normalize parameters, as gradient descent works much better if
// the function being optimized is homogeneous in direction
// for example f = x^2 + y^2 is easier to optimize than f = x^2 + 50*y^2
```

It's my assumption here that we would want the parameter weight for the `y` dimension (the second weight) to be half as much as the first weight (for `x`) because the function contains `2Y^2`.

So something like `GradientDescentOptimizer<2> gradientDescentOptimizer({0.1, 0.05});`

Would that be more correct or is my understanding flawed here?

Reply:
Great point, and totally correct understanding. Changed both instances of this to be:
```
    // Note that we halve the weight for "y" here to make sure
    // gradient descent can see the function in a more homogenous way.
    // See the GradientDescentOptimizer class javadoc comment for more details.
    GradientDescentOptimizer<2> gradientDescentOptimizer({0.1, 0.05});
```